controllers/git_tekton_resources_renovater.go

```diff
@@ -0,0 +1,268 @@
+		dependencyDashboard: false
+	}
+	`
+	return fmt.Sprintf(template, slug, slug)
+}
+
+func (r *GitTektonResourcesRenovater) CreateRenovaterJob(ctx context.Context, installations []github.ApplicationInstallation) error {
+
+	name := fmt.Sprintf("renovate-job-%s", getRandomString(5))
+	secretTokens := map[string]string{}
+	renovateCmds := []string{}
+	for _, installation := range installations {
+		secretTokens[fmt.Sprint(installation.InstallationID)] = installation.Token
+		renovateCmds = append(renovateCmds, fmt.Sprintf("RENOVATE_TOKEN=$TOKEN_%d renovate", installation.InstallationID))
+	}
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: buildServiceNamespaceName,
+		},
+		StringData: secretTokens,
+	}
+	trueBool := true
+	falseBool := false
+	backoffLimit := int32(1)
+	timeToLive := int32(TimeToLiveOfJob.Seconds())
+	job := &batch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: buildServiceNamespaceName,
+		},
+		Spec: batch.JobSpec{
+			BackoffLimit:            &backoffLimit,
+			TTLSecondsAfterFinished: &timeToLive,
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Volumes: []corev1.Volume{
+						{
+							Name: RenovateConfigName,
+							VolumeSource: corev1.VolumeSource{
+								ConfigMap: &corev1.ConfigMapVolumeSource{
+									LocalObjectReference: corev1.LocalObjectReference{Name: RenovateConfigName},
+								},
+							},
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name:  "renovate",
+							Image: RenovateImageUrl,
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									Prefix: "TOKEN_",
+									SecretRef: &corev1.SecretEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: name,
+										},
+									},
+								},
+							},
+							Command: []string{"bash", "-c", strings.Join(renovateCmds, "; ")},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      RenovateConfigName,
+									MountPath: "/usr/src/app/config.js",
+									SubPath:   "config.js",
+								},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Capabilities:             &corev1.Capabilities{Drop: []corev1.Capability{"ALL"}},
+								RunAsNonRoot:             &trueBool,
+								AllowPrivilegeEscalation: &falseBool,
+								SeccompProfile: &corev1.SeccompProfile{
+									Type: corev1.SeccompProfileTypeRuntimeDefault,
+								},
+							},
+						},
+					},
+					RestartPolicy: corev1.RestartPolicyNever,
+				},
+			},
+		},
+	}
+
+	if err := r.Client.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	if err := r.Client.Delete(ctx, secret); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	if err := r.Client.Create(ctx, secret); err != nil {
+		return err
+	}
+	if err := r.Client.Create(ctx, job); err != nil {
```

Comment:
What if deletion in background will not be completed when this line executed?

Reply:
Yeah, this may happen to the Delete & Create pair for the ConfigMap, Secret and CronJob. If we really want to manage such objects programmatically in controller, and avoid potential conflict issue happening in k8s, it maybe worth considering append a random string suffix or something else that could make each object as unique in the cluster.

Reply:
added timestamp to job name and it's also contain random string, so I removed the deletion. Deletetion of old job is handled by  TTLSecondsAfterFinished on job and secret is linked to the job.